django_emarsys/config.py

```diff
@@ -0,0 +1,160 @@
+from django.conf import settings
+from django.core.checks import Critical, Error, Warning
+
+
+_VALID_NAME = re.compile(r'^[a-zA-Z0-9_ ]+$')
+_VALID_ARG = re.compile(r'^[a-z][a-z0-9_]*$')
+
+
+def _make_message_set(messages):
+    """
+    Unfortunately `django.core.checks.CheckMessage` doesn't implement
+    `__hash__()`, so sets containing them cannot be compared directly.
+    As a workaround this makes a set containing tuples of the level
+    and message, which are the only two attributes the tests care about.
+
+    Example:
+    >>> from django.core.checks import Critical
+    >>> c1 = Critical("foobar")
+    >>> c2 = Critical("foobar")
+    >>> c1 == c2
+    True
+    >>> {c1} == {c2}
+    False
+    >>> _make_message_set({c1}) == _make_message_set({c2})
+    True
+
+    """
+    return {(m.level, m.msg) for m in messages}
+
+
+class AssertCheckMessagesEqual:
+    def assertCheckMessagesEqual(self, l1, l2):
+        self.assertSetEqual(_make_message_set(l1), _make_message_set(l2))
+
+
+def validate_settings():
+    """
+    Turn settings into a configuration dict that `validate_configuration` can
+    check.
+
+    The extra `hasattr()` check is necessary to make this work with overridden
+    settings from tests. If a setting existed in the original settings object,
+    but is then deleted in overridden settings, then `dir()` will report it
+    present, but accessing it raises an `AttributeError`.
+
+    Minimal test for this:
+
+    >>> from django.conf import settings
+    >>> from django.test.utils import override_settings
+    >>> @override_settings()
+    >>> def test_strange_dir_issue():
+    >>>     delattr(settings, 'DEBUG')
+    >>>     assert 'DEBUG' in dir(settings)
+    >>>     assert not hasattr(settings, 'DEBUG')
+    >>>
+    >>> test_strange_dir_issue()
+
+    """
+    return validate_configuration({key: getattr(settings, key)
+                                  for key in dir(settings)
+                                  if hasattr(settings, key)})
+
+
+def flatten(l):
+    """
+    Flatten a list of lists into a list
+    """
+    return list(chain.from_iterable(l))
+
+
+def duplicates(l):
+    """
+    Return non-unique elements in a list.
+
+    Should be in Python.
+    """
+    return [item for item, count in Counter(l).items() if count > 1]
+
+
+def validate_configuration(config):
+    messages = []
+    if not config.get('EMARSYS_ACCOUNT'):
+        messages.append(Critical("EMARSYS_ACCOUNT not set"))
+
+    if not config.get('EMARSYS_PASSWORD'):
+        messages.append(Critical("EMARSYS_PASSWORD not set"))
+
+    if not config.get('EMARSYS_BASE_URI'):
+        messages.append(Critical("EMARSYS_BASE_URI not set"))
+
+    events = config.get('EMARSYS_EVENTS')
+    if not events:
+        events = {}
+        messages.append(Critical("EMARSYS_EVENTS not set"))
+
+    messages += flatten([_validate_event(event, params)
+                         for event, params
+                         in events.items()])
```

Comment:
Can probably get rid of the `[]` here - flatten should iterate over a generator expression just fine, like in line 126 in `_validate_event`.